aten/src/ATen/native/TensorTransformations.cpp

```diff
@@ -13,80 +13,132 @@ namespace native {
+    : num_indexers(num_indexers)
+    , indexers(indexers)
+    , indexer_strides(indexer_strides) {
+
+  }
+
+  int64_t num_indexers;
+  char** indexers;
+  const int64_t* indexer_strides;
+
+  int64_t get(int64_t idx) {
+    int64_t offset = *(int64_t*)&indexers[0][idx * indexer_strides[0]];
+    for (int j = 1; j < num_indexers; j++) {
+      offset += *(int64_t*)&indexers[j][idx * indexer_strides[j]];
+      // offset += value;
     }
-  });
+    return offset;
+  }
+};
+
+Tensor build_index(int64_t num_dims, int64_t flip_dim, int64_t dim_size) {
+  auto new_shape = std::vector<int64_t>(num_dims, 1);
+  new_shape[flip_dim] = dim_size;
+
+  TensorOptions tensor_options =
+    TensorOptions(c10::kLong).
+    device(c10::kCPU);
+
+  return at::empty(new_shape, tensor_options);
 }
 
-Tensor flip_cpu(const Tensor& self, IntArrayRef dims) {
-  auto in_tensor = self;
-  const int64_t total_dims = in_tensor.dim();
-  auto flip_dims_b = at::dim_list_to_bitset(dims, total_dims);
-  Tensor out_tensor = at::empty_like(in_tensor, LEGACY_CONTIGUOUS_MEMORY_FORMAT);
-
-  // create contiguous strides for input tensor
-  auto stride_contiguous_v = std::vector<int64_t>(total_dims);
-  for (int64_t i = total_dims - 1; i >= 0; i--) {
-    if (i == total_dims - 1) {
-      stride_contiguous_v[i] = 1;
-    } else {
-      stride_contiguous_v[i] = std::max<int64_t>(in_tensor.size(i + 1), 1) * stride_contiguous_v[i + 1];
+
+std::vector<Tensor> build_indices_loop(Tensor input, IntArrayRef flip_dims) {
+  std::vector<Tensor> indices;
+  int64_t element_size_bytes = input.element_size();
+  for(auto dim: flip_dims) {
+    auto dim_size = input.size(dim);
+    auto index = build_index(input.ndimension(), dim, dim_size);
+    auto stride = input.stride(dim);
+    auto input_index_ptr = index.data_ptr<int64_t>();
+
+    for(int64_t i = 0; i < dim_size; i++) {
+      input_index_ptr[i] = static_cast<int64_t>(dim_size - i - 1) * stride * element_size_bytes;
     }
+    indices.push_back(index);
+  }
+  return indices;
+}
+
+static TensorIterator make_index_iterator(const Tensor input, const std::vector<Tensor> indices) {
+  TensorIteratorConfig config;
+  config.set_check_mem_overlap(false)
+        .check_all_same_dtype(false)
+        .declare_static_dtype_and_device(input.scalar_type(), input.device())
+        .add_output(Tensor())
+        .add_input(input);
+  for (auto& index : indices) {
+    config.add_input(index);
   }
+  return config.build();
+}
+
+template<typename scalar_t>
+void flip_cpu_kernel(TensorIterator& iter) {
+  int ntensor = iter.ntensors();
+  // When launch the index parallel version, set a relative samll grain size less than the INTERNAL::GRAIN_SIZE
+  // to make the whole available thread numbers get more balanced work load and a better cache location.
+  // The grain size here is chosen by the op benchmark to overcome the thread launch overhead
+  const int index_parallel_grain_size = 3000;
```

Comment:
Did we actually benchmark this number? If not, then the comment might be stale. This was taken from the advanced indexing kernel, so it might not apply here. Maybe just point out that we are using the same constant as in the advanced indexing code?

Reply:
No, we didn't benchmark this number! I'll put the corresponding comment there